module/state_synchronization/indexer/execution.go

```diff
@@ -0,0 +1,182 @@
+// If the register was not found the storage.ErrNotFound error is returned.
+func (i *ExecutionState) RegisterValues(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	first, err := i.registers.FirstHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read first index height: %w", err)
+	}
+
+	last, err := i.registers.LatestHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read last index height: %w", err)
+	}
+
+	if height < first || height > last {
+		return nil, fmt.Errorf("register out of indexed height bounds, current height range: [%d, %d], requested height: %d", first, last, height)
+	}
+
+	values := make([]flow.RegisterValue, len(IDs))
+
+	for j, id := range IDs {
+		value, err := i.registers.Get(id, height)
+		if err != nil {
+			return nil, err
+		}
+
+		values[j] = value
+	}
+
+	return values, nil
+}
+
+// IndexBlockData indexes all execution block data by height.
+// If the height was already indexed the operation will be ignored.
+func (i *ExecutionState) IndexBlockData(ctx context.Context, data *execution_data.BlockExecutionDataEntity) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	block, err := i.headers.ByBlockID(data.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not get the block by ID %s: %w", data.BlockID, err)
+	}
+
+	// progress height in storage first, if the height is incorrect the index storage will panic
+	// this is due to unrecoverable bug in the implementation
+	err = i.registers.SetLatestHeight(block.Height)
+	if err != nil {
+		return fmt.Errorf("failed to progress index height to %d: %w", block.Height, err)
+	}
+
+	// concurrently process indexing of block data
+	g, ctx := errgroup.WithContext(ctx)
+	for j, chunk := range data.ChunkExecutionDatas {
+		g.Go(func() error {
+			err := i.indexEvents(data.BlockID, chunk.Events)
+			if err != nil {
+				return fmt.Errorf("could not index events for chunk %d: %w", j, err)
+			}
+			return nil
+		})
+		g.Go(func() error {
+			err = i.indexCommitment(flow.StateCommitment(chunk.TrieUpdate.RootHash), block.Height)
+			if err != nil {
+				return fmt.Errorf("could not index events for chunk %d: %w", j, err)
+			}
+			return nil
+		})
+		g.Go(func() error {
+			err = i.indexRegisterPayload(chunk.TrieUpdate.Payloads, block.Height)
+			if err != nil {
+				return fmt.Errorf("could not index registers for chunk %d: %w", j, err)
+			}
+			return nil
+		})
+	}
+
+	err = g.Wait()
+	if err != nil {
+		return fmt.Errorf("failed to index block data at height %d: %w", block.Height, err)
+	}
+
+	return nil
+}
+
+func (i *ExecutionState) indexCommitment(commitment flow.StateCommitment, height uint64) error {
+	i.commitments[height] = commitment
+	return nil
+}
+
+func (i *ExecutionState) indexEvents(blockID flow.Identifier, events flow.EventsList) error {
+	// Note: service events are currently not included in execution data: https://github.com/onflow/flow-go/issues/4624
```

Comment:
Note: the last chunk in an execution data is the system chunk. All events in that `ChunkExecutionData` are service events.